sklearn/utils/validation.py

```diff
@@ -498,6 +500,11 @@ def check_array(array, accept_sparse=False, accept_large_sparse=True,
         estimator_name = "Estimator"
     context = " by %s" % estimator_name if estimator is not None else ""
 
+    # When all dataframe columns are sparse, convert to a sparse array
+    if hasattr(array, 'sparse') and array.ndim > 1:
```

Comment:
To be safe, I would also add a check that all columns are indeed sparse (`sparse` now raises if not all columns are sparse, but I am not sure this is necessarily guaranteed to stay that way). 
The explicit check would be something like:

```
df.dtypes.apply(pd.api.types.is_sparse).all()
```

See also https://github.com/pandas-dev/pandas/issues/26706

Reply:
I moved my comment to a new issue: https://github.com/scikit-learn/scikit-learn/issues/16845